extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,491 @@
+	/**
+	 * Access token
+	 */
+	token: string;
+
+	/**
+	 * TokenType
+	 */
+	tokenType: string;
+}
+
+export interface TokenClaims { // https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+	aud: string;
+	iss: string;
+	iat: number;
+	idp: string,
+	nbf: number;
+	exp: number;
+	c_hash: string;
+	at_hash: string;
+	aio: string;
+	preferred_username: string;
+	email: string;
+	name: string;
+	nonce: string;
+	oid: string;
+	roles: string[];
+	rh: string;
+	sub: string;
+	tid: string;
+	unique_name: string;
+	uti: string;
+	ver: string;
+}
+
+export type TokenRefreshResponse = { accessToken: AccessToken, refreshToken: RefreshToken, tokenClaims: TokenClaims };
+
+export abstract class AzureAuth {
+	protected readonly memdb = new MemoryDatabase();
+
+	protected readonly WorkSchoolAccountType: string = 'work_school';
+	protected readonly MicrosoftAccountType: string = 'microsoft';
+
+	protected readonly loginEndpointUrl: string;
+	protected readonly commonTenant: string;
+	protected readonly redirectUri: string;
+	protected readonly scopes: string[];
+	protected readonly scopesString: string;
+	protected readonly clientId: string;
+	protected readonly resources: Resource[];
+
+
+	constructor(
+		protected readonly metadata: AzureAccountProviderMetadata,
+		protected readonly tokenCache: SimpleTokenCache,
+		protected readonly context: vscode.ExtensionContext,
+		protected readonly authType: AzureAuthType,
+		public readonly userFriendlyName: string
+	) {
+		this.loginEndpointUrl = this.metadata.settings.host;
+		this.commonTenant = 'common';
+		this.redirectUri = this.metadata.settings.redirectUri;
+		this.clientId = this.metadata.settings.clientId;
+
+		this.resources = [
+			this.metadata.settings.armResource, this.metadata.settings.sqlResource,
+			this.metadata.settings.graphResource, this.metadata.settings.ossRdbmsResource
+		];
+
+		this.scopes = [...this.metadata.settings.scopes];
+		this.scopesString = this.scopes.join(' ');
+	}
+
+	public abstract async login(): Promise<AzureAccount | azdata.PromptFailedResult>;
+
+	public abstract async autoOAuthCancelled(): Promise<void>;
+
+
+	public async refreshAccess(account: azdata.Account): Promise<azdata.Account> {
+		const response = await this.getCachedToken(account.key);
+		if (!response) {
+			account.isStale = true;
+			return account;
+		}
+
+		const refreshToken = response.refreshToken;
+		if (!refreshToken || !refreshToken.key) {
+			account.isStale = true;
+			return account;
+		}
+
+		try {
+			await this.refreshAccessToken(account.key, refreshToken);
+		} catch (ex) {
+			if (ex.message) {
+				await vscode.window.showErrorMessage(ex.message);
+			}
+			console.log(ex);
+		}
+		return account;
+	}
+
+
+	public async getSecurityToken(account: azdata.Account, azureResource: azdata.AzureResource): Promise<TokenResponse> {
+		const resource = this.resources.find(s => s.azureResourceId === azureResource);
+		if (!resource) {
+			return undefined;
```

Comment:
Would be good to update return type and make sure callers are handling undefined correctly

Reply:
Yep! I'll try to potentially enable strict null checks for azurecore too